### PR TITLE
Fix ambiguity parsing collection in conditional expressions

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/75318

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@cston @RikkiGibson ptal :)

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs [hunk 1]
```diff
@@ -2206,41 +2206,42 @@ public void NullableType_Is_01()
         public void NullableType_Is_02()
         {
             string source = "x is string ? [A] y";
-            UsingExpression(source);
+            UsingExpression(source,
+                // (1,19): error CS1003: Syntax error, ':' expected
+                // x is string ? [A] y
+                Diagnostic(ErrorCode.ERR_SyntaxError, "y").WithArguments(":").WithLocation(1, 19));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs:2212]: note: this change is appropriate.  it's not legal to say `x is string?[A]`.  `string?[A]` is *not* a type.  While you can instantiate a `new string?[A]`, it's type is now `string?[]`.  So this now properly becomes a case where this is a conditional that creates a collection, but is missing a colon.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs [hunk 2]
```diff
@@ -2305,10 +2306,7 @@ public void NullableType_Is_03()
         public void NullableType_Is_04()
         {
             string source = "_ = x is string ? [A] y => y : z";
-            UsingExpression(source,
-                // (1,1): error CS1073: Unexpected token '=>'
-                // _ = x is string ? [A] y => y : z
-                Diagnostic(ErrorCode.ERR_UnexpectedToken, "_ = x is string ? [A] y").WithArguments("=>").WithLocation(1, 1));
+            UsingExpression(source);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs:2309]: the old parsing was just wrong.  This should have parsed according to the rules of the language.  The precedence rules say: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1115-conditional-operator

```g4
conditional_expression
    : null_coalescing_expression
    | null_coalescing_expression '?' expression ':' expression
    ;
```

So the true/false parts of a conditional can be full expressions (which includes lambdas).  So this should have correctly parsed before.  This is a fix.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs [hunk 3]
```diff
@@ -2360,13 +2368,7 @@ public void NullableType_Is_04()
         public void NullableType_Is_05()
         {
             string source = "_ = x is string ? [return: A] y => y : z";
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs:2370]: same here.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 4]
```diff
@@ -7367,10 +7367,43 @@ bool canBeNullableType()
...
+                                    {
+                                        // `is T?[,]`.  Definitely an array of nullable type.
+                                        case SyntaxKind.CommaToken:
+                                            return true;
+
+                                        // `is T?[]`.  Could be an array of a nullable type, or a conditional.  Have to
+                                        // see if it is followed by `:` to find out.  If there is a colon, it's a
+                                        // conditional.
+                                        case SyntaxKind.CloseBracketToken:
+                                            return this.PeekToken(3).Kind != SyntaxKind.ColonToken;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7392]: If collection-exprs become usable as a primary-expression, we'll have to revisit this heuristic, right?
e.g. for `x is T ? [].Method() : ...`

Also, are there any potentially valid expression forms today, which don't have the collection immediately followed by `:`, like `x is T ? [] == GetCollection() : ...`?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 5]
```diff
@@ -11254,13 +11287,16 @@ private ExpressionSyntax ParseExpressionContinued(ExpressionSyntax leftOperand,
                 }
             }
 
-            // From the language spec:
+            // https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md?rgh-link-date=2024-10-02T19%3A11%3A19Z#1115-conditional-operator:
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:11290]: Does the `rgh-link-date=...` in this url do something useful? If not, consider removing it.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7392]: > If collection-exprs become usable as a primary-expression, we'll have to revisit this heuristic, right?

Yes.  We'd need to revisit that based on what can legally follow.   Fun times.

> Also, are there any potentially valid expression forms today, which don't have the collection immediately followed by :, like x is T ? [] == GetCollection() : ...?

THis is definitely an interesting case.  I'm thinking we may actually have to speculatively parse here to see what is happening.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: ick.  will do.

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 6]
```diff
@@ -7351,33 +7351,17 @@ private TypeSyntax ParseTypeCore(ParseTypeMode mode)
             {
                 switch (this.CurrentToken.Kind)
                 {
-                    case SyntaxKind.QuestionToken when canBeNullableType():
+                    case SyntaxKind.QuestionToken:
                         {
-                            var question = EatNullableQualifierIfApplicable(mode);
+                            var question = TryEatNullableQualifierIfApplicable(type, mode);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7356]: ok.  the prior logic was very confusing.  We had 'canBeNullable' here, and then EatNullableQualifierIfApplicable (which also did a bunch *more* canBeNullable checks).  I've merged these all into the renamed TryEatNullableQualifierIfApplicable method now.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -7351,33 +7351,17 @@ private TypeSyntax ParseTypeCore(ParseTypeMode mode)
...
-                            goto done; // token not consumed
-                        }
-                        bool canBeNullableType()
-                        {
-                            // These are the fast tests for (in)applicability.
-                            // More expensive tests are in `EatNullableQualifierIfApplicable`
-                            if (type.Kind == SyntaxKind.NullableType || type.Kind == SyntaxKind.PointerType)
-                                return false;
-                            if (this.PeekToken(1).Kind == SyntaxKind.OpenBracketToken)
-                                return true;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7371]: this is the logic that had to be expanded on majorly to handle the new parsing.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -7351,33 +7351,17 @@ private TypeSyntax ParseTypeCore(ParseTypeMode mode)
...
-                            if (mode == ParseTypeMode.DefinitePattern)
-                                return true; // Permit nullable type parsing and report while binding for a better error message
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7373]: don't need this check anymore, as this was jsut the bool that determined if we moved onto TryEatNullableQualifierIfApplicable

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 9]
```diff
@@ -7351,33 +7351,17 @@ private TypeSyntax ParseTypeCore(ParseTypeMode mode)
...
-                            if (mode == ParseTypeMode.NewExpression && type.Kind == SyntaxKind.TupleType &&
-                                this.PeekToken(1).Kind is not SyntaxKind.OpenParenToken and not SyntaxKind.OpenBraceToken)
-                            {
-                                return false; // Permit `new (int, int)?(t)` (creation) and `new (int, int) ? x : y` (conditional)
-                            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7378]: this falls out from existing code in the existing method.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 10]
```diff
@@ -7417,32 +7403,81 @@ bool canBeNullableType()
...
         {
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.QuestionToken);
-            using var resetPoint = this.GetDisposableResetPoint(resetOnDispose: false);
+
+            // These are the fast tests for (in)applicability. More expensive tests are follow.
+            //
+            // If we already have `x?` or `x*` then do not parse out a nullable type if we see `x??` or `x*?`.  These
+            // are never legal as types in the language, so we can fast bail out.
+            if (typeParsedSoFar.Kind is SyntaxKind.NullableType or SyntaxKind.PointerType)
+                return null;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7428]: this was a move, and i added comments to explain it.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 11]
```diff
@@ -7417,32 +7403,81 @@ bool canBeNullableType()
...
-                resetPoint.Reset();
+                outerResetPoint.Reset();
                 return null;
             }
 
             return questionToken;
 
             bool canFollowNullableType()
             {
+                if (mode == ParseTypeMode.AfterIs && this.CurrentToken.Kind is SyntaxKind.OpenBracketToken)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7444]: this is the new logic for figuring out what to do with `x?[`

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 12]
```diff
@@ -7417,32 +7403,81 @@ bool canBeNullableType()
...
+                        case SyntaxKind.CloseBracketToken:
+                            {
+                                using var innerResetPoint = this.GetDisposableResetPoint(resetOnDispose: true);
+
+                                // Consume the expression after the `?`.
+                                var whenTrue = this.ParsePossibleRefExpression();
+
+                                // Now see if we have a ':' following.  If so, this is a conditional.  If not, it's a nullable type.
+                                return this.CurrentToken.Kind != SyntaxKind.ColonToken;
+                            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7472]: this was added based on what Rikki noticed.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 13]
```diff
@@ -7523,7 +7558,10 @@ bool canFollowNullableType()
                         // If nothing from above worked permit the nullable qualifier if it is followed by a token that
                         // could not start an expression. If we have `T?[]` we do want to treat that as an array of
                         // nullables (following existing parsing), not a conditional that returns a list.
-                        return !CanStartExpression() || this.CurrentToken.Kind is SyntaxKind.OpenBracketToken;
+                        if (this.CurrentToken.Kind is SyntaxKind.OpenBracketToken)
+                            return true;
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7562]: pulled simple check up.

File: src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs [hunk 14]
```diff
@@ -2202,58 +2202,59 @@ public void NullableType_Is_01()
...
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75318")]
         public void NullableType_Is_03()
         {
             string source = "_ = x is string ? [] y => y : z";
             UsingExpression(source,
-                // (1,1): error CS1073: Unexpected token '=>'
+                // (1,20): error CS1001: Identifier expected
                 // _ = x is string ? [] y => y : z
-                Diagnostic(ErrorCode.ERR_UnexpectedToken, "_ = x is string ? [] y").WithArguments("=>").WithLocation(1, 1));
+                Diagnostic(ErrorCode.ERR_IdentifierExpected, "]").WithLocation(1, 20));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/LambdaAttributeParsingTests.cs:2257]: this parsing matches teh parsing of hte test below this one (`"_ = x is string ? [A] y => y : z"`).  Specifically, we now see this *appropriately* as a conditional that returns a lambda.  But we report an issue on the lambda that it's attribute-list is missing attributes.  This is  better parse than before, more in line with the parse once you add the attribute.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 15]
```diff
@@ -7417,32 +7403,81 @@ bool canBeNullableType()
...
+                        // `is T?[,]`.  Definitely an array of nullable type.
+                        case SyntaxKind.CommaToken:
+                            return true;
+
+                        // `is T?[]`.  Could be an array of a nullable type, or a conditional.  Have to
+                        // see if it is followed by `:` to find out.  If there is a colon, it's a
+                        // conditional.
+                        case SyntaxKind.CloseBracketToken:
+                            {
+                                using var innerResetPoint = this.GetDisposableResetPoint(resetOnDispose: true);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7465]: nit: `using var _` here?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:7465]: sure!

**Review by 333fred — APPROVED** [review]

**Pull request merged**